crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -41,22 +41,22 @@ use crate::Db;
 
 mod except_handlers;
 
-/// Are we in a state where a `break` statement is allowed?
-#[derive(Clone, Copy, Debug)]
-enum LoopState {
-    InLoop,
-    NotInLoop,
+#[derive(Clone, Debug, Default)]
+struct Loop {
+    /// Flow states at each `break` in the current loop.
+    break_states: Vec<FlowSnapshot>,
```

Comment:
I haven't looked through all usages but I'm wondering if we could use a single `break_states` struct stored on `SemanticIndexBuilder` instead of allocating a `Vec` for each `Loop`. 

What I'm thinking about:

* `Loop`: stores the length of the "arena" at the time this loop was pushed
* Getting all break states for the current loop is the same as `arena[start..]`
* When popping a loop, truncate the arena to `start` to remove all snapshots belonging to the current state.

Reply:
I had originally thought I might invest more time in reducing vec allocations here, but the simple version was completely neutral on benchmarks, so I'm not sure it's worth putting more time into this right now. I think perf optimization work should be motivated by some evidence.

I suspect that in practice loops are relatively uncommon, and loops with `break` in them even more uncommon, so in the great majority of cases we never allocate for this vec at all.